src/Stratis.Bitcoin/P2P/Peer/NetworkPeerClient.cs

```diff
@@ -0,0 +1,172 @@
+﻿using System;
+using System.IO;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.P2P.Peer
+{
+    /// <summary>
+    /// Represents a TCP client that can connect to a server and send and receive messages.
+    /// </summary>
+    public class NetworkPeerClient : IDisposable
+    {
+        /// <summary>Logger factory to create loggers.</summary>
+        private readonly ILoggerFactory loggerFactory;
+
+        /// <summary>Instance logger.</summary>
+        private ILogger logger;
+
+        /// <summary>Unique identifier of a client.</summary>
+        public int Id { get; private set; }
+
+        /// <summary>Underlaying TCP client.</summary>
+        private TcpClient tcpClient;
+
+        /// <summary>Prevents parallel execution of multiple write operations on <see cref="Stream"/>.</summary>
+        private AsyncLock writeLock;
```

Comment:
I find `writeLock` a bit confusing, this is a plane lock and there is not read lock

Reply:
it is lock to protect write operations to the stream, therefore writeLock, i think it makes perfect sense :) there is no read lock because reading is done in serial fashion, but writing can be requested by multiple threads at the time